IdentityCore/src/requests/broker/MSIDSSOExtensionGetSsoCookiesRequest.m

```diff
@@ -0,0 +1,99 @@
+
+@property (nonatomic, copy) MSIDGetSsoCookiesRequestCompletionBlock requestCompletionBlock;
+
+@end
+
+@implementation MSIDSSOExtensionGetSsoCookiesRequest
+
+- (nullable instancetype)initWithRequestParameters:(MSIDRequestParameters *)requestParameters
+                                 accountIdentifier:(MSIDAccountIdentifier *)accountIdentifier
+                                            ssoUrl:(NSString *)ssoUrl
+                                     correlationId:(NSUUID *)correlationId
+                                             error:(NSError **)error{
+    self = [super initWithRequestParameters:requestParameters error:error];
+    if (self)
+    {
+        _accountIdentifier = accountIdentifier;
+        _ssoUrl = ssoUrl;
+        _correlationId = correlationId;
+        
+        __typeof__(self) __weak weakSelf = self;
+        self.extensionDelegate.completionBlock = ^(MSIDBrokerNativeAppOperationResponse *operationResponse, NSError *error)
+        {
+            NSArray *prtHeaders = nil;
+            NSArray *deviceHeaders = nil;
+            NSError *resultError = error;
+            
+            if (!operationResponse.success)
+            {
+                MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, requestParameters, @"Finished get sso cookies request with error %@", MSID_PII_LOG_MASKABLE(error));
+            }
+            else if (![operationResponse isKindOfClass:[MSIDBrokerOperationGetSsoCookiesResponse class]])
+            {
+                resultError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Received incorrect response type for the get sso cookies request", nil, nil, nil, nil, nil, YES);
+            }
+            else
+            {
+                MSIDBrokerOperationGetSsoCookiesResponse *response = (MSIDBrokerOperationGetSsoCookiesResponse *)operationResponse;
+                prtHeaders = response.prtHeaders;
+                deviceHeaders = response.deviceHeaders;
+            }
+            
+            MSIDGetSsoCookiesRequestCompletionBlock completionBlock = weakSelf.requestCompletionBlock;
+            weakSelf.requestCompletionBlock = nil;
+            
+            if (completionBlock) completionBlock(prtHeaders, deviceHeaders, resultError);
+        };
+        
+        self.ssoProvider = [ASAuthorizationSingleSignOnProvider msidSharedProvider];
+    }
+    
+    return self;
+}
+
+- (void)executeRequestWithCompletion:(nonnull MSIDGetSsoCookiesRequestCompletionBlock)completionBlock
+{
+    MSIDBrokerOperationGetSsoCookiesRequest *getSsoCookiesRequest = [MSIDBrokerOperationGetSsoCookiesRequest new];
+    getSsoCookiesRequest.accountIdentifier = self.accountIdentifier;
+    getSsoCookiesRequest.ssoUrl = self.ssoUrl;
+    getSsoCookiesRequest.correlationId = self.correlationId ?: [NSUUID UUID];
+    
+    __typeof__(self) __weak weakSelf = self;
+    [self executeBrokerOperationRequest:getSsoCookiesRequest continueBlock:^{
+        weakSelf.requestCompletionBlock = completionBlock;
```

Comment:
Why are we setting this after executing? Wouldn't there be a race condition?

Reply:
The continue block/error block worked on the same thread in sync. So there should not be race condition. 
This was set in the following order:

```    [self.authorizationController performRequests];```
```    self.requestCompletionBlock = completionBlock;```

Since I call the base class and try to reuse performRequest, and the completionBlock is in different types in different Sso request. That is why I send the next code execution within the block